pubtools/_impl/tracing.py

```diff
@@ -0,0 +1,145 @@
+import threading
+
+from opentelemetry import baggage, context, trace
+from opentelemetry.baggage.propagation import W3CBaggagePropagator
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import (BatchSpanProcessor,
+                                            ConsoleSpanExporter)
+from opentelemetry.trace import Status, StatusCode
+from opentelemetry.trace.propagation.tracecontext import \
+    TraceContextTextMapPropagator
+
+from pubtools.pluggy import pm
+
+propagator = TraceContextTextMapPropagator()
+baggage_propagator = W3CBaggagePropagator()
+TRACE_WRAPPER = None
+log = logging.getLogger(__name__)
+
+
+def get_trace_wrapper():
+    """return a global trace wrapper instance"""
+    global TRACE_WRAPPER
+    if TRACE_WRAPPER is None:
+        TRACE_WRAPPER = TracingWrapper()
+    return TRACE_WRAPPER
+
+
+class TracingWrapper:
+    """Wrapper class to initialize opentelemetry instrumentation and provide a helper function
+    for instrumenting a function"""
+
+    def __init__(self):
+        self._processor = None
+        self._provider = None
+        if os.getenv("OTEL_TRACING", "").lower() == "true":
+            log.info("Creating TracingWrapper instance")
+            exporter = pm.hook.otel_exporter() or ConsoleSpanExporter()
+            self._processor = BatchSpanProcessor(exporter)
+            self._provider = TracerProvider(
+                resource=Resource.create({SERVICE_NAME: os.getenv("OTEL_SERVICE_NAME")})
+            )
+            self._provider.add_span_processor(self._processor)
+            trace.set_tracer_provider(self._provider)
+            set_global_textmap(propagator)
+
+    def instrument_func(self, span_name=None, carrier=None, args_to_attr=False):
+        """Instrument tracing for a function.
+
+        Args:
+            span_name: str
+                Span name. It's assigned with the function's name by default if it's omitted.
+            carrier: dict
+                A dictionary which holds trace context. Trace context will be extracted from it if
+                if it's provided.
+            args_to_attr: boolean
+                Add function parameters into span attributes or not.
+
+        Returns:
+            The decorated function
+        """
+        tracer = trace.get_tracer(__name__)
+
+        def _instrument_func(func):
+            @functools.wraps(func)
+            def wrap(*args, **kwargs):
+                attributes = {
+                    "function_name": func.__qualname__,
+                }
+                if args_to_attr:
+                    attributes["args"] = ", ".join(map(str, args))
+                    attributes["kwargs"] = ", ".join(
+                        "{}={}".format(k, v) for k, v in kwargs.items()
+                    )
+
+                if not os.getenv("OTEL_TRACING", "").lower() == "true":
```

Comment:
Minor: it's probably a good idea to do this env var check only once, in the TracingWrapper constructor, and store the result as `self._enabled` or similar.

Otherwise we have to worry about what happens if you have the env var unset, construct a TracingWrapper, instrument functions, then set the env var and call the instrumented functions.  It'll be kind of half enabled, half disabled at that point.

Reply:
Fixed it.